File: L/interview47-li-wu-de-zui-da-jie-zhi-lcof/main.go
package main

import "fmt"

func maxValue(grid [][]int) int {
	buf := make([][]int, len(grid))
	for i := 0; i < len(grid); i++ {
		buf[i] = make([]int, len(grid[0]))
	}

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			up, left := 0, 0
			if i > 0 {
				up = buf[i-1][j]
			}
			if j > 0 {
				left = buf[i][j-1]
			}
			sum := up
			if up < left {
				sum = left
			}
			buf[i][j] = sum + grid[i][j]
		}
	}
	return buf[len(grid)-1][len(grid[0])-1]
}

func main() {
	fmt.Println("a")
}

File: L/interview47-li-wu-de-zui-da-jie-zhi-lcof/main.rs
struct Solution {}

impl Solution {
  pub fn max_value(grid: Vec<Vec<i32>>) -> i32 {
    let mut buf: Vec<Vec<i32>> = vec![vec![0; grid[0].len()]; grid.len()];

    (0..grid.len()).for_each(|r| {
      (0..grid[0].len()).for_each(|c| {
        let mut up: i32 = 0;
        let mut left: i32 = 0;
        if r > 0 {
          up = buf[r - 1][c];
        }
        if c > 0 {
          left = buf[r][c - 1];
        }
        let mut inter: i32 = up;
        if inter < left {
          inter = left;
        }
        buf[r][c] = inter + grid[r][c];
      });
    });

    buf[buf.len() - 1][buf[0].len() - 1]
  }
}
